src/runtime/page_manager.cc

```diff
@@ -0,0 +1,224 @@
+void BlockAllocator::free(PhysicalTokenBlock& block) {
+    if (block.ref_count == 0) {
+        printf("block is already freed\n");
+        throw std::runtime_error("Double free! Block is already freed.");
+    }
+    block.decr_ref_count();
+    if (block.ref_count == 0) {
+        // printf("put block number: %d back to free_blocks\n", block.get_block_number());
+        free_blocks.push_back(block);
+    }else{
+        // in current implementation this should not be the case
+        printf("block is not freed. Ref count: %d\n", block.ref_count);
+        throw std::runtime_error("Block is not freed. Ref count: " + std::to_string(block.ref_count));
+    }
+}
+
+int BlockAllocator::get_num_free_blocks() const {
+    return free_blocks.size();
+}
+
+PageManager::PageManager(int block_size, int num_total_blocks)
+    : block_size(block_size), num_total_blocks(num_total_blocks),
+      block_allocator(block_size, num_total_blocks) {
+      }
+
+//return the physical number of this block
+int PageManager::allocate_one_block(const RequestGuid& request_guid) {
+    BlockTable& block_table = block_tables[request_guid];
+
+    PhysicalTokenBlock block = block_allocator.allocate();
+    block_table.push_back(block);
+    block_tables[request_guid] = block_table;
+    return block.get_block_number();
+}
+
+void PageManager::free_block_table(BlockTable& block_table) {
+    // make it reverse order to free the last allocated block first
+    BlockTable::reverse_iterator rit = block_table.rbegin();
+    for (; rit != block_table.rend(); ++rit) {
+        block_allocator.free(*rit);
+    }
+    return;
+}
+
+void PageManager::free_request(const RequestGuid& request_guid) {
+    //we only free the blocks that are already used
+    assert(block_tables.find(request_guid) != block_tables.end());
+    BlockTable block_table = block_tables[request_guid];
+    free_block_table(block_table);
+    block_tables.erase(request_guid);
+    return;
+}
+
+// delete the last num_blocks in the request_guid
+void PageManager::free_multiple_blocks(const RequestGuid& request_guid, int num_blocks) {
+    assert(block_tables.find(request_guid) != block_tables.end());
+    auto& block_table = block_tables[request_guid];
+    assert(num_blocks <= block_table.size());
+    int num_blocks_allocated = block_table.size();
+    for (int i = 0; i < num_blocks; i++) {
+        block_allocator.free(block_table[num_blocks_allocated - i - 1]);
+    }
+    // only keep the first num_blocks_allocated - num_blocks blocks
+    block_table.erase(block_table.begin() + num_blocks_allocated - num_blocks, block_table.end());
+    block_tables[request_guid] = block_table;
+    return;
+}
+
+// int PageManager::get_index_last_block(const RequestGuid& request_guid) const {
+//     const auto& block_table = block_tables.at(request_guid);
+//     return block_table.back.get_block_number();
+// }
+
+std::vector<int> PageManager::get_block_table_indices(const RequestGuid& request_guid) const {
```

Comment:
It looks not quite effient to compute block indices every calling time, especially given long context where we may have hundreds of block for each request.

Reply:
Thanks Zhuofu. After some thoughts I think this would be the best way to do it. As I have one extra level of abstraction for the index (i.e., physical block), I have to unfold the indices every time. Though it is O(n), it will only be used once and since each block contains 64 tokens, 2000 tokens will only use up to 30 pages so I think it won't be too expensive.